src/Database/LSMTree/Internal.hs

```diff
@@ -450,6 +457,42 @@ close th = RW.modify_ (tableHandleState th) $ \case
         V.forM_ residentRuns $ Run.removeReference (tableHasFS thEnv)
       pure TableHandleClosed
 
+{-# SPECIALISE lookups :: V.Vector SerialisedKey -> TableHandle IO h -> (Maybe (Entry SerialisedValue (BlobRef (Run (Handle h)))) -> lookupResult) -> IO (V.Vector lookupResult) #-}
+-- | See 'Database.LSMTree.Normal.lookups'.
+lookups ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => V.Vector SerialisedKey
+  -> TableHandle m h
+  -> (Maybe (Entry SerialisedValue (BlobRef (Run (Handle h)))) -> lookupResult)
+     -- ^ How to map from an entry to a lookup result. Use
+     -- 'toNormalLookupResult' or 'toMonoidalLookupResult'.
+  -> m (V.Vector lookupResult)
+lookups ks th fromEntry = withOpenTable th $ \thEnv -> do
+    wb <- readMVar (tableWriteBuffer thEnv)
+    let !wbRes = WB.lookups' wb ks
+    let resolve = maybe const getResolveMupsert (confResolveMupsert $ tableConfig thEnv)
+    cache <- readMVar (tableCache thEnv)
+    ioRes <-
+      lookupsInBatches
+        (tableHasBlockIO thEnv)
+        (BatchSize $ ioctxBatchSizeLimit (FS.getParams (tableHasBlockIO thEnv)))
+        resolve
+        (cachedRuns cache)
+        (cachedFilters cache)
+        (cachedIndexes cache)
+        (cachedKOpsFiles cache)
+        ks
+    pure $! V.zipWith (\e1 e2 -> fromEntry $ combineMaybe resolve e1 e2) wbRes ioRes
```

Comment:
This all looks pretty reasonable now.

```suggestion
    pure $! V.zipWith (\e1 e2 -> fromEntry $! combineMaybe resolve e1 e2) wbRes ioRes
```

Reply:
I think `V.zipWith` is probably not strict enough itself. I've changed it to:

```suggestion
    pure $! V.zipWithStrict (\e1 e2 -> fromEntry $ combineMaybe resolve e1 e2) wbRes ioRes
```

Reply:
I think you're right.